Refresh token

forslund opened the pull request:
The "oauth.refresh" message is now handled and will trigger a token
refresh if possible. The Message data must contain valid "skill_id" and
"app_id" fields.

It also adds a basic "expires_at" entry to the token data to help determining if a token needs to be renewed.

forslund committed:
Add "expires_at" to token data if applicable

If the expires_in key exists the expires_at will be added. This can be
useful to determine if the token needs refreshing without a failing call
to the api.

The expires_at is an absolute time when the token expires.

fixup

forslund committed:
Add support for refreshing oauth token

The "oauth.refresh" message is now handled and will trigger a token
refresh if possible. The Message data must contain valid "skill_id" and
"app_id" fields.

forslund committed:
gitignore: Add vim cache files

NeonDaniel commented on ovos_PHAL_plugin_oauth/__init__.py:
```diff
@@ -58,6 +59,10 @@ def oauth_callback(munged_id):
         ).json()
 
     with OAuthTokenDatabase() as db:
+        if 'expires_at' in token_response:
```
Looks like this should be `if 'expires_in' in token_response`?

emphasize replied:
I'm puzzling about this. Google is sending the key `expires_at` (UTC). nvm, got it.

forslund replied:
I barely remember this code myself. My guess is that it was a quick and dirty way to get a simple local time or it was something spotify-specific I was doing.

emphasize replied:
just checked googles token fetch response has it both (1 hour token)
```
{'access_token': 'xxx', 'expires_in': 3599, 'scope': ['https://www.googleapis.com/auth/calendar'], 'token_type': 'Bearer', 'expires_at': 1704829227.0002847}
```

forslund replied:
After checking a bit I believe the spotify response only has the `expires_in`. So it should probably be changed to be

```python
 if 'expires_at' not in token_response:
    token_response['expires_at'] = time.time() + token_response['expires_in']
```

emphasize replied:
this also has to/should be reflected at
```diff
+ 224            new_token_data['expires_at'] = time.time() + token_data['expires_in']
```

NeonDaniel commented on ovos_PHAL_plugin_oauth/__init__.py:
```diff
@@ -182,6 +188,47 @@ def handle_oauth_register(self, message):
+        """
+        oauth_id = f"{message.data['skill_id']}_{message.data['app_id']}"
+        # Load all needed data for refresh
+        with OAuthApplicationDatabase() as db:
```
After a rebase on `dev`, there's a class variable for this so it doesn't need to be instantiated on every reference

NeonDaniel commented on ovos_PHAL_plugin_oauth/__init__.py:
```diff
@@ -182,6 +188,47 @@ def handle_oauth_register(self, message):
+                token_data.update(new_token_data)
+                db.update_token(oauth_id, token_data)
+        else:
+            LOG.error("Token refresh failed :(")
```
Consider emitting a response message so the caller can be notified of success/failure

emphasize replied:
Also to consider, we already have a msg type in use returning the response from an initial fetch
https://github.com/OpenVoiceOS/ovos-PHAL-plugin-oauth/blob/126455896f0ff66c518404641299ce55887289d6/ovos_PHAL_plugin_oauth/__init__.py#L64-L66

forslund replied:
Is "oauth.token.response..." a good message name for the refresh result? Seems like it could be cause for confusion?

NeonDaniel reviewed (changes requested):
Left some feedback, only one thing looks like a bug/typo but left some other suggestions to consider.

forslund commented:
Thanks for the feedback, I'll fix the things mentioned and update the PR

emphasize commented on ovos_PHAL_plugin_oauth/__init__.py:
```diff
@@ -182,6 +188,47 @@ def handle_oauth_register(self, message):
+            return
+
+        refresh_token = token_data['refresh_token']
+        refresh_endpoint = app_data["refresh_endpoint"]
```
Maybe changing this to 

`refresh_endpoint = app_data.get("refresh_endpoint") or app_data["token_endpoint"]`

as at least google with the newer API versions uses the token endpoint for both.

forslund replied:
Yes, that might be a bit more failsafe. For spotify I set token endpoint as refresh endpoint during initial setup so I assumed it would be left to the user/dev to set it up properly.

forslund replied:
Reading up a bit on oauth2 again. The refresh endpoint was copied from spotipy's implementation but according to https://auth0.com/docs/authenticate/protocols/oauth it should always be the token endpoint. So I think the refresh_endpoint should be removed altogether from the db as required argument...I'll make an issue over at the backend for further discussion.

forslund commented:
I've done some updates according to discussion above: updated the expires_at criteria, added response to the incoming refresh request. (If consensus is that oauth.token.response.xxx should be used I can change that).

Is there a standard for how OVOS handles "bad" results that I should use? now I've added a result field with a string (Error/Ok)

Edit: Also thanks for all the feedback!
